# BUILDING A REGRESSION MODEL IN KERAS

In [14]:
# Import the necessary libraries and explore the data set

import pandas as pd
import numpy as np
import keras

from keras.models import Sequential
from keras.layers import Dense

In [15]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')

In [16]:
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [17]:
# X are predictors and y is the outcome (splitting the dataset into predictors and target/outcome)

X = concrete_data.drop(columns=["Strength"])
y = concrete_data["Strength"]
X.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [18]:
y.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [19]:
n_cols = X.shape[1]

# PART A - Model creation, training and evaluation

In [20]:
def regression_model():
    model = Sequential()
    model.add(Dense(10, activation = 'relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    model.compile(optimizer = 'adam', loss = 'mean_squared_error')
    return model

In [21]:
model = regression_model()

In [22]:
# Data are split into train and test sets

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=3)

In [23]:
model.fit(X_train, y_train, epochs=50)

Epoch 1/50
721/721 [==============================] - 1s 716us/step - loss: 234927.0606
Epoch 2/50
721/721 [==============================] - 0s 303us/step - loss: 84246.3167
Epoch 3/50
721/721 [==============================] - 0s 301us/step - loss: 22684.6225
Epoch 4/50
721/721 [==============================] - 0s 225us/step - loss: 8425.9475
Epoch 5/50
721/721 [==============================] - 0s 306us/step - loss: 6671.9016
Epoch 6/50
721/721 [==============================] - 0s 278us/step - loss: 6285.0809
Epoch 7/50
721/721 [==============================] - 0s 470us/step - loss: 5641.8038
Epoch 8/50
721/721 [==============================] - 0s 276us/step - loss: 5007.6141
Epoch 9/50
721/721 [==============================] - 0s 254us/step - loss: 4483.8509
Epoch 10/50
721/721 [==============================] - 0s 288us/step - loss: 4022.1695
Epoch 11/50
721/721 [==============================] - 0s 323us/step - loss: 3628.3695
Epoch 12/50
721/721 [===========================

In [24]:
y_pred = model.predict(X_test)
from sklearn.metrics import mean_squared_error
print("Mean squared error on the test data is %.3f" %(mean_squared_error(y_test, y_pred)))

Mean squared error on the test data is 172.083


# Repeated steps 1-3 50 times using a for-loop
## With means

In [25]:
n_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, 50):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = i)
    model.fit(X_train, y_train, epochs = epochs, verbose = 0)
    MsqE = model.evaluate(X_test, y_test, verbose = 0)
    print("MsqE" +str(i + 1)+": "+str(MsqE))
    y_pred = model.predict(X_test)
    MSE = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(MSE)
    print('.', end='')
print('End!')

MsqE1: 97.45278012173847
.MsqE2: 125.51388426351701
.MsqE3: 108.26858540260291
.MsqE4: 121.31393872109817
.MsqE5: 132.92334567073092
.MsqE6: 123.0597766271301
.MsqE7: 134.2143368520397
.MsqE8: 110.45137018987661
.MsqE9: 138.53962218645708
.MsqE10: 110.0513467140568
.MsqE11: 104.37951931752819
.MsqE12: 98.48648212025466
.MsqE13: 81.16211267736738
.MsqE14: 99.48390029857845
.MsqE15: 53.54519695294328
.MsqE16: 52.12736571722432
.MsqE17: 49.4908289492709
.MsqE18: 56.71992209968443
.MsqE19: 46.743236405949766
.MsqE20: 53.26810603465849
.MsqE21: 44.76849483749241
.MsqE22: 47.137735706317
.MsqE23: 55.09545534251191
.MsqE24: 53.61364974481774
.MsqE25: 49.29740425529604
.MsqE26: 60.07863720643868
.MsqE27: 73.04058637896787
.MsqE28: 64.78380235344846
.MsqE29: 54.252942353776355
.MsqE30: 52.00625714052071
.MsqE31: 54.385240190623264
.MsqE32: 50.91816250946144
.MsqE33: 48.7551111758334
.MsqE34: 50.729433917690635
.MsqE35: 62.30237576876643
.MsqE36: 56.26368972630177
.MsqE37: 58.55486088971875
.Msq

In [26]:
mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)
print("The mean squared errors of 50 regression models: mean = %.3f, standard deviation = %.3f" %(mean, standard_deviation))

The mean squared errors of 50 regression models: mean = 71.153, standard deviation = 28.072


# Part B - Normalizing the data and repeating part A steps

In [28]:
X_norm = (X - X.mean()) / X.std()
X_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [29]:
n_cols = X_norm.shape[1]
n_cols

8

# Creating the regression model using Keras

In [30]:
def regression_model():
    model = Sequential()
    model.add(Dense(10, activation = 'relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    # Compile the model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [31]:
# Splitting the data into train and test sets

X_norm_train, X_norm_test, y_train, y_test = train_test_split(X_norm, y, test_size=0.3, random_state=3)

# Training and evaluating the network

In [33]:
model = regression_model()
model.fit(X_norm_train, y_train, epochs = 50)

Epoch 1/50
721/721 [==============================] - 1s 853us/step - loss: 1547.5741
Epoch 2/50
721/721 [==============================] - 0s 309us/step - loss: 1531.2023
Epoch 3/50
721/721 [==============================] - 0s 355us/step - loss: 1515.1619
Epoch 4/50
721/721 [==============================] - 0s 300us/step - loss: 1499.2823
Epoch 5/50
721/721 [==============================] - 0s 332us/step - loss: 1483.5743
Epoch 6/50
721/721 [==============================] - 0s 280us/step - loss: 1467.6576
Epoch 7/50
721/721 [==============================] - 0s 308us/step - loss: 1451.6607
Epoch 8/50
721/721 [==============================] - 0s 305us/step - loss: 1435.0997
Epoch 9/50
721/721 [==============================] - 0s 358us/step - loss: 1418.3394
Epoch 10/50
721/721 [==============================] - 0s 331us/step - loss: 1400.8685
Epoch 11/50
721/721 [==============================] - 0s 636us/step - loss: 1382.7954
Epoch 12/50
721/721 [==============================]

In [35]:
y_pred = model.predict(X_norm_test)
from sklearn.metrics import mean_squared_error
print("Mean squared error on the test data is %.3f" %(mean_squared_error(y_test, y_pred)))

Mean squared error on the test data is 457.688


# Repeated steps 1-3 50 times using a for-loop
## With means

In [40]:
n_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, 50):
    X_norm_train, X_norm_test, y_train, y_test = train_test_split(X_norm, y, test_size = 0.3, random_state = i)
    model.fit(X_norm_train, y_train, epochs = epochs, verbose = 0)
    MsE = model.evaluate(X_norm_test, y_test, verbose = 0)
    print("MsE" +str(i + 1)+": "+str(MsE))
    y_pred = model.predict(X_norm_test)
    MSE = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(MSE)
    print('.', end='')
print('End!')

MsE1: 95.18161658870363
.MsE2: 118.43436634347663
.MsE3: 104.83406121213845
.MsE4: 114.41845117957847
.MsE5: 111.14443391966589
.MsE6: 96.08330069236385
.MsE7: 106.55319169424112
.MsE8: 78.38717021756959
.MsE9: 85.12673824273267
.MsE10: 74.71829038453333
.MsE11: 69.19914280172304
.MsE12: 63.65630520817531
.MsE13: 77.33818318388609
.MsE14: 78.35761396784612
.MsE15: 68.29302615563846
.MsE16: 64.91700502661054
.MsE17: 66.40280724189043
.MsE18: 68.32060477494421
.MsE19: 62.6908458968968
.MsE20: 71.36093901353361
.MsE21: 59.18636650406427
.MsE22: 71.92019280492295
.MsE23: 58.45836225836794
.MsE24: 52.882749674775454
.MsE25: 54.218514501084016
.MsE26: 55.70596444336728
.MsE27: 48.75875785358515
.MsE28: 54.34585272532836
.MsE29: 49.87145752891368
.MsE30: 44.12826382535175
.MsE31: 43.32750615487207
.MsE32: 41.76928735010832
.MsE33: 43.80153283177842
.MsE34: 42.88184829514389
.MsE35: 42.26286614674195
.MsE36: 49.85460073971054
.MsE37: 40.23496526576169
.MsE38: 43.6721770894566
.MsE39: 39.019380

In [41]:
mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)
print("The mean squared errors of 50 regression models: mean = %.3f, standard deviation = %.3f" %(mean, standard_deviation))

The mean squared errors of 50 regression models: mean = 61.139, standard deviation = 22.631


#### QUESTION: How does the mean of the mean squared errors compared to that from step A? The mean of 61.139 is lower than the mean of 71.153 from step A.

## PART C - Repeating Part B, but using 100 epochs for training

In [42]:
model = regression_model()
model.fit(X_norm_train, y_train, epochs = 100)

Epoch 1/100
721/721 [==============================] - 1s 834us/step - loss: 1615.9509
Epoch 2/100
721/721 [==============================] - 0s 328us/step - loss: 1598.7875
Epoch 3/100
721/721 [==============================] - 0s 302us/step - loss: 1581.7893
Epoch 4/100
721/721 [==============================] - 0s 300us/step - loss: 1564.3656
Epoch 5/100
721/721 [==============================] - 0s 306us/step - loss: 1547.0319
Epoch 6/100
721/721 [==============================] - 0s 395us/step - loss: 1529.2113
Epoch 7/100
721/721 [==============================] - 0s 325us/step - loss: 1510.6158
Epoch 8/100
721/721 [==============================] - 0s 283us/step - loss: 1491.3412
Epoch 9/100
721/721 [==============================] - 0s 330us/step - loss: 1471.4669
Epoch 10/100
721/721 [==============================] - 0s 326us/step - loss: 1449.7744
Epoch 11/100
721/721 [==============================] - 0s 331us/step - loss: 1427.7274
Epoch 12/100
721/721 [===================

# Repeated steps 1-3 50 times using a for-loop
## With means

In [44]:
n_mean_squared_errors = 50
epochs = 100
mean_squared_errors = []
for i in range(0,50):
    X_norm_train, X_norm_test, y_train, y_test = train_test_split(X_norm, y, test_size = 0.3, random_state = i)
    model.fit(X_norm_train, y_train, epochs = epochs, verbose = 0)
    MsE = model.evaluate(X_norm_test, y_test, verbose = 0)
    print("MsE" +str(i + 1)+": "+str(MsE))
    y_pred = model.predict(X_norm_test)
    MSE = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(MSE)
    print('.', end='')
print('End!')

MsE1: 28.20871966710754
.MsE2: 28.701945554862903
.MsE3: 26.85003960865601
.MsE4: 28.754442116203432
.MsE5: 29.173058889444594
.MsE6: 29.063732807690272
.MsE7: 32.18633444332382
.MsE8: 26.094706421145343
.MsE9: 29.821905413877616
.MsE10: 24.43673788768188
.MsE11: 29.096370968618054
.MsE12: 28.698902802945728
.MsE13: 28.921507032943776
.MsE14: 32.62849337383381
.MsE15: 30.1164937189096
.MsE16: 24.43087262397445
.MsE17: 32.06820150949422
.MsE18: 29.678994039887364
.MsE19: 27.15353046651797
.MsE20: 30.559525153397743
.MsE21: 26.46144816327635
.MsE22: 26.954262576056916
.MsE23: 25.829328759202678
.MsE24: 27.732937235662465
.MsE25: 27.638846147407605
.MsE26: 27.689808700462763
.MsE27: 26.732618115866455
.MsE28: 26.09132799438674
.MsE29: 31.351051157732226
.MsE30: 27.854092446731517
.MsE31: 25.772869005172385
.MsE32: 25.64353966944426
.MsE33: 24.46673530282326
.MsE34: 28.327345888205716
.MsE35: 31.544678981250158
.MsE36: 32.86586992331693
.MsE37: 25.815123530267513
.MsE38: 28.602311192978547

In [45]:
mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)
print("The mean squared errors of 50 regression models: mean = %.3f, standard deviation = %.3f" %(mean, standard_deviation))

The mean squared errors of 50 regression models: mean = 28.266, standard deviation = 2.225


#### QUESTION: How does the mean of the mean squared errors compared to that from step B? The mean of 28.266 is much lower than the mean of 61.139 from step B.

## PART D - Repeating Part B using 3 hidden layers of 10 neurons each

In [46]:
def regression_model():
    model = Sequential()
    model.add(Dense(10, activation = 'relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation = 'relu'))
    model.add(Dense(10, activation = 'relu'))
    model.add(Dense(1))
    # Compile the model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [47]:
# Splitting the data into train and test sets

X_norm_train, X_norm_test, y_train, y_test = train_test_split(X_norm, y, test_size=0.3, random_state=3)

# Training and evaluating the network

In [48]:
model = regression_model()
model.fit(X_norm_train, y_train, epochs = 50)

Epoch 1/50
721/721 [==============================] - 1s 1ms/step - loss: 1567.3720
Epoch 2/50
721/721 [==============================] - 0s 555us/step - loss: 1538.4464
Epoch 3/50
721/721 [==============================] - 0s 475us/step - loss: 1508.3951
Epoch 4/50
721/721 [==============================] - 0s 439us/step - loss: 1469.2859
Epoch 5/50
721/721 [==============================] - 0s 468us/step - loss: 1412.9785
Epoch 6/50
721/721 [==============================] - 0s 470us/step - loss: 1332.9684
Epoch 7/50
721/721 [==============================] - 0s 499us/step - loss: 1222.4522
Epoch 8/50
721/721 [==============================] - 0s 499us/step - loss: 1067.9127
Epoch 9/50
721/721 [==============================] - 0s 447us/step - loss: 883.5476
Epoch 10/50
721/721 [==============================] - 0s 412us/step - loss: 675.1400
Epoch 11/50
721/721 [==============================] - 0s 423us/step - loss: 486.7313
Epoch 12/50
721/721 [==============================] - 0s

# Repeated steps 1-3 50 times using a for-loop
## With means

In [49]:
n_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, 50):
    X_norm_train, X_norm_test, y_train, y_test = train_test_split(X_norm, y, test_size = 0.3, random_state = i)
    model.fit(X_norm_train, y_train, epochs = epochs, verbose = 0)
    MsE = model.evaluate(X_norm_test, y_test, verbose = 0)
    print("MsE" +str(i + 1)+": "+str(MsE))
    y_pred = model.predict(X_norm_test)
    MSE = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(MSE)
    print('.', end='')
print('End!')

MsE1: 94.68186575386517
.MsE2: 59.994988907502304
.MsE3: 36.55982035417773
.MsE4: 37.86330196541104
.MsE5: 34.9611834060027
.MsE6: 33.32375674571806
.MsE7: 35.293529029031404
.MsE8: 26.12970684261384
.MsE9: 28.877981043942153
.MsE10: 25.448458933907418
.MsE11: 26.269907972959253
.MsE12: 21.59439160985854
.MsE13: 29.083642539854573
.MsE14: 28.57559936177769
.MsE15: 27.284211272946454
.MsE16: 19.258243866337157
.MsE17: 22.510994179734904
.MsE18: 25.95815007354835
.MsE19: 24.659487616282835
.MsE20: 26.797577842539567
.MsE21: 25.963753388537558
.MsE22: 24.450119993833276
.MsE23: 21.300723714735902
.MsE24: 23.368823116265453
.MsE25: 24.230691669056718
.MsE26: 25.34922137461048
.MsE27: 22.400633747137867
.MsE28: 23.367122718045625
.MsE29: 24.666892795501017
.MsE30: 23.744628856868804
.MsE31: 19.703332413361682
.MsE32: 23.48962896155694
.MsE33: 20.52670159694832
.MsE34: 22.634982383752718
.MsE35: 25.068516987427152
.MsE36: 24.23908734244436
.MsE37: 18.55989226554204
.MsE38: 23.196796071567967

In [50]:
mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)
print("The mean squared errors of 50 regression models: mean = %.3f, standard deviation = %.3f" %(mean, standard_deviation))

The mean squared errors of 50 regression models: mean = 27.098, standard deviation = 11.645


#### QUESTION: How does the mean of the mean squared errors compared to that from step B? The mean of 27.098 is much lower than the mean of 61.139 from step B.